tensorflow/core/kernels/stack.cc

```diff
@@ -231,52 +313,64 @@ void StackPushOp::ComputeAsync(OpKernelContext* ctx, DoneCallback done) {
     return;
   }
 
-  // Push the tensor onto the stack. Swap the tensor to CPU if instructed.
+  // Push the tensor onto the stack.
+  // Swap the oldest tensor to CPU if instructed.
   const Tensor& tensor = ctx->input(1);
   AllocatorAttributes alloc_attrs = ctx->input_alloc_attr(1);
+
+  OP_REQUIRES_OK_ASYNC(ctx, stack->Push({tensor, alloc_attrs, false}), done);
+  ctx->set_output(0, tensor);
+
   // For now, we use a simple heuristic for swapping: A GPU tensor is moved
   // to CPU if the tensor has more than kCopyThreshold bytes and the GPU
   // allocator says more than kOccupancy of the memory is in use.
   static constexpr int kCopyThreshold = 2048;
   static constexpr double kOccupancy = 0.7;
-  if (swap_memory_ && !alloc_attrs.on_host() &&
```

Comment:
This may seem like a big change, but actually most of it is converting from nested ifs to early returns. An observation shows this is true.

The reason for early returns is so that `done()` is called as late as possible.